core-backend/common/src/memory.rs

```diff
@@ -0,0 +1,214 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Work with WASM program memory in backends.
+
+use core::{marker::PhantomData, mem::size_of};
+
+use alloc::vec::Vec;
+use codec::{Decode, DecodeAll, MaxEncodedLen};
+use gear_core::{
+    buffer::{RuntimeBuffer, RuntimeBufferSizeError},
+    env::Ext,
+    memory::Memory,
+};
+use gear_core_errors::MemoryError;
+
+#[derive(Debug, derive_more::Display, derive_more::From)]
+pub enum MemoryAccessError {
+    #[from]
+    #[display(fmt = "{_0}")]
+    Memory(MemoryError),
+    #[from]
+    #[display(fmt = "{_0}")]
+    RuntimeBuffer(RuntimeBufferSizeError),
+    DecodeError,
+    #[display(fmt = "Buffer size {_0} is not eq to pre-registered size {_1}")]
+    WrongBufferSize(usize, u32),
+}
+
+/// Memory access manager. Allows to pre-register memory accesses,
+/// and pre-process, them together. For example:
+/// ```ignore
+/// let manager = MemoryAccessManager::default();
+/// let read1 = manager.new_read(10, 20);
+/// let read2 = manager.new_read_as::<u128>(100);
+/// let write1 = manager.new_write_as::<usize>(190);
+///
+/// // First call of read or write interface leads to pre-processing of
+/// // all already registered memory accesses, and clear `self.reads` and `self.writes`.
+/// let value_u128 = manager.read_as(read2).unwrap();
+///
+/// // Next calls do not lead to access pre-processing.
+/// let value1 = manager.read().unwrap();
+/// manager.write_as(write1, 111).unwrap();
+/// ```
+#[derive(Debug)]
+pub struct MemoryAccessManager<E: Ext> {
+    reads: Vec<(u32, u32)>,
+    writes: Vec<(u32, u32)>,
```

Comment:
I would introduce two new types for better readability and self-documenting source code: `Pointer` and `Size`.

Reply:
Done